cpp/src/Slice/DeprecationReporter.cpp

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) ZeroC, Inc.
+
+#include "DeprecationReporter.h"
+
+#include <sstream>
+
+using namespace std;
+using namespace Slice;
+
+namespace
+{
+    class DeprecationVisitor final : public ParserVisitor
```

Comment:
Unfortunately, a visitor is the only way to check this, because of how compilation works in Ice. Bison parses from the inside out, so we don't know what metadata will be applied to something until we've already fully parsed it.